### PR TITLE
Show callstack in browser upon REPL crashes

### PR DESCRIPTION
**Pull request by caiismyname**
Fixing some incompatibility between heapGraph and `__optimize()`. 
Essence of the issue was an implicit assumption that only one visitor would visit certain entries. 

Original Issue: https://github.com/facebook/prepack/issues/1732

### CODE REVIEW

File: src/utils/generator.js [hunk 1]
```diff
@@ -253,10 +253,11 @@ class ModifiedBindingEntry extends GeneratorEntry {
       containingGenerator === this.containingGenerator,
       "This entry requires effects to be applied and may not be moved"
     );
-    invariant(this.modifiedBinding.value === this.newValue);
+    invariant(this.modifiedBinding.value === this.newValue, "ModifiedBinding's value has been improperly modified.");
```
Comment (hermanventer) [inline src/utils/generator.js]: This is a good example of an utterly useless comment. Much better (and much harder) is to write a comment that explains to a reader of the code why this invariant is expected to hold. In terms of debug logging, for those who care about such things, it would be better to include information about the two values.

File: scripts/test-runner.js [hunk 2]
```diff
@@ -286,6 +286,7 @@ function runTest(name, code, options: PrepackOptions, args) {
   options.invariantLevel = code.includes("// omit invariants") || args.verbose ? 0 : 99;
   if (code.includes("// emit concrete model")) options.emitConcreteModel = true;
   if (code.includes("// exceeds stack limit")) options.maxStackDepth = 10;
+  if (code.includes("// heapGraphFilePath")) options.heapGraphFormat = "VISJS";
```
Comment (yinghuitan) [inline scripts/test-runner.js]: It is not a file path, you are using VISJS format here. I would call "// heap graph" instead of `heapGraphFilePath `

File: src/serializer/serializer.js [hunk 3]
```diff
@@ -182,6 +182,9 @@ export class Serializer {
             additionalFunctionValuesAndEffects,
             referentializer
           );
+          // Copying dERB map so cache lookup returns the same object, and the subsequent
```
Comment (yinghuitan) [inline src/serializer/serializer.js]: I would write the comment like: ModifiedBindingEntry caches the state from declarativeEnvironmentRecordsBindings during visitor so we have to pass the same cache to future visitors.

File: scripts/test-runner.js (hunk 2 above)
Comment (caiismyname) [reply scripts/test-runner.js]: Will do. I put it like that originally to avoid confusion with the actual flag (the flag itself is `--heapGraphFilePath` so I kept it the same)

Comment (yinghuitan) [reply scripts/test-runner.js]: Yeah, `heapGraphFilePath` is a command line level option not actually the engine/serializer level option, which is `heapGraphFormat`. But either way, we are just trying to enable heap graph and do not care about the output in this test.

File: src/utils/generator.js [hunk 4]
```diff
@@ -253,10 +253,17 @@ class ModifiedBindingEntry extends GeneratorEntry {
...
+      "ModifiedBinding's value has been changed since last visit."
+    );
     let [residualBinding, newValue] = context.visitModifiedBinding(this.modifiedBinding);
-    invariant(this.residualFunctionBinding === undefined || this.residualFunctionBinding === residualBinding);
+    invariant(
+      this.residualFunctionBinding === undefined || this.residualFunctionBinding === residualBinding,
+      "ResidualFunctionBinding has been changed since last visit."
+    );
     this.residualFunctionBinding = residualBinding;
+    this.modifiedBinding.value = newValue;
```
Comment (NTillmann) [inline src/utils/generator.js]: Is this change really needed? Looks a bit scary to me. I'd rather do less state mutation on the side than more.

Comment (caiismyname) [reply src/utils/generator.js]: Without it, we run into the follow problem:

In the provided example
```
(function () {
  let a = __abstract("number", "a");
  let b, c;
  global.f = function() {
    b = a + 42;
    c = a + 42;
  }
  __optimize(f);
})();
```
when the `ModifiedBindingEntry` does the `.visitModifiedBinding`, there is a `.visitEquivalentValue(newValue)` call on the new value, and the returned newValue may not match the original (it's replaced by the "equivalent" value if one is found). This happens to us in this test case because of the 
```
b = a + 42;
c = a + 42;
```
since the `a + 42` look the same. The assignment realigns the state to what is expected by the invariant assertion. 

This problem actually seems to be at least somewhat independent of the problem caused by the cache/ResidualFunctionBinding, but it was in the same code area and also causes the code to crash. 

It seems that the more fundamental issue is that we're re-traversing a the heap, which has some sort of assumption that it won't be re-traversed.

Comment (NTillmann) [reply src/utils/generator.js]: > which has some sort of assumption that it won't be re-traversed.

Yes, that's the underlying problem here...

While the introduced assignment makes the invariant violation go away, it changes what happens during visiting, which is bad. 

The right thing to do would be to stop mutating state on the side in these GeneratorEntry's, they should be just values. Instead, build up new data structure in the visitor, a mapping of GeneratorEntry to something --- whatever mutable state there was, and pass that on to the next phase that really needs it, and just throw it away after the `ResidualHeapGraphVisitor`.

Comment (caiismyname) [reply src/utils/generator.js]: @yinghuitan was saying then that it might be easier to detect and reject calls to `__optimize()` when the `--heapGraphFilePath` flag is given, since the two haven't been made to work together yet. Would that be a more relevant solution?

File: src/serializer/serializer.js [hunk 5]
```diff
@@ -175,18 +175,24 @@ export class Serializer {
         );
 
         if (this.options.heapGraphFormat) {
+          if (additionalFunctionValuesAndEffects.size > 0) {
+            let diagnostic = new CompilerDiagnostic(
+              "Using __optimize with --heapGraphFilePath flag is not currently supported",
+              undefined,
```
Comment (caiismyname) [inline src/serializer/serializer.js]: Wasn't sure what to give these fields since they're not really applicable. Suggestions?

**Comment by caiismyname** [comment]
Screenshot of new behavior on the code snippet that previously caused the REPL to crash

File: scripts/test-runner.js [hunk 6]
```diff
@@ -286,6 +286,7 @@ function runTest(name, code, options: PrepackOptions, args) {
   options.invariantLevel = code.includes("// omit invariants") || args.verbose ? 0 : 99;
   if (code.includes("// emit concrete model")) options.emitConcreteModel = true;
   if (code.includes("// exceeds stack limit")) options.maxStackDepth = 10;
+  if (code.includes("// heapGraph")) options.heapGraphFormat = "VISJS";
```
Comment (yinghuitan) [inline scripts/test-runner.js]: I do not think we need this change anymore.
